cylc/uiserver/authorise.py

```diff
@@ -0,0 +1,477 @@
+        # We won't be re-checking auth for return variables
+        if len(info.path) > 1:
+            return next_(root, info, **args)
+        op_name = self.get_op_name(info.field_name, info.operation.operation)
+        # It shouldn't get here but worth checking for zero trust
+        if not op_name:
+            self.auth_failed(self.current_user, op_name, 400,
+                             "Operation not in schema.")
+        authorised = self.auth.is_permitted(self.current_user, op_name)
+        if not authorised:
+            self.auth_failed(self.current_user, op_name, 403)
+        if (info.operation.operation in Authorization.ASYNC_OPS
+                or iscoroutinefunction(next_)):
+            return self.async_resolve(next_, root, info, **args)
+        return next_(root, info, **args)
+
+    @staticmethod
+    def auth_failed(current_user: str, op_name: str,
+                    http_code: int, message: Optional[str] = None):
+        """
+        Raise authorization error
+        Args:
+            current_user: username accessing operation
+            op_name: operation name
+            http_code: http error code to raise
+            message: Message to log Defaults to None.
+
+        Raises:
+            web.HTTPError
+        """
+        log_message = (f"Authorization failed for {current_user}"
+                       f":requested to {op_name}.")
+        if message:
+            log_message = log_message + " " + message
+        logger.warn(log_message)
+        raise web.HTTPError(http_code)
+
+    @staticmethod
+    def get_op_name(field_name: str, operation: str) -> Union[None, str]:
+        """
+        Returns operation name required for authorization.
+        Converts queries and subscriptions to read operations.
+        Args:
+            field_name: Field name e.g. play
+            operation: operation type
+
+        Returns:
+            operation name
+        """
+        if operation in Authorization.READ_AUTH_OPS:
+            return Authorization.READ_OPERATION
+        else:
+            # Check it is a mutation in our schema
+            if (field_name in Authorization.ALL_OPS):
+                return field_name
+        return None
+
+    async def async_resolve(self, next_, root, info, **args):
+        """Return awaited coroutine"""
+        return await next_(root, info, **args)
+
+
+def get_groups(username: str) -> List[str]:
+    """Return list of system groups for given user
+
+    Args:
+        username: username used to check system groups.
+
+    Returns:
+        list: system groups for username given
+    """
+    groupmax = (os.NGROUPS_MAX)  # type: ignore
+    group_ids = os.getgrouplist(username, groupmax)
+    group_ids.remove(groupmax)
+    # turn list of group_ids into group names with group identifier prepended
+    return list(map(
+        lambda x: f'{Authorization.GRP_IDENTIFIER}{grp.getgrgid(x).gr_name}',
+        group_ids
+        )
+    )
```

Comment:
Note to self: test with that LDAP library, to confirm we can get the users too (NIWA uses it and I think we have a really old issue in cylc-flow related to that...).

Reply:
Excellent, would be great to test the groups as thoroughly as we can.

Reply:
I've mocked and created a non-existing user. Anything that returns the user string name should work, including the SSSD/LDAP created users :+1: marking as resolved.